internal/venice-consumer/src/main/java/com/linkedin/venice/client/consumer/VeniceChangeLogConsumerImpl.java

```diff
@@ -0,0 +1,288 @@
+              pubSubTopicPartition,
+              offset,
+              timestamp,
+              payloadSize);
+          pubSubMessages.add(pubSubMessage);
+        }
+      }
+    }
+    return pubSubMessages;
+  }
+
+  private PubSubMessage<K, V, Long> convertRecordToPubSubMessage(
+      KafkaMessageEnvelope kafkaMessageEnvelope,
+      K currentKey,
+      PubSubTopicPartition pubSubTopicPartition,
+      Long offset,
+      Long timestamp,
+      int payloadSize) {
+    MessageType messageType = MessageType.valueOf(kafkaMessageEnvelope.messageType);
+    V currentValue = null;
+    switch (messageType) {
+      case PUT:
+        Put put = (Put) kafkaMessageEnvelope.payloadUnion;
+        Schema valueSchema = schemaReader.getValueSchema(put.schemaId);
+        RecordDeserializer<V> valueDeserializer =
+            FastSerializerDeserializerFactory.getFastAvroGenericDeserializer(valueSchema, valueSchema);
+        currentValue = valueDeserializer.deserialize(put.putValue);
+        break;
+    }
+    return new ImmutablePubSubMessage(currentKey, currentValue, pubSubTopicPartition, offset, timestamp, payloadSize);
+  }
+
+  private PubSubMessage<K, ChangeEvent<V>, Long> convertChangeEventToPubSubMessage(
+      Put put,
+      K currentKey,
+      PubSubTopicPartition pubSubTopicPartition,
+      Long offset,
+      Long timestamp,
+      int payloadSize) {
+    byte[] valueBytes = put.putValue.array();
+    RecordChangeEvent recordChangeEvent = recordChangeDeserializer.deserialize(valueBytes);
+    Schema currentValueSchema = schemaReader.getValueSchema(1);
+    RecordDeserializer<V> valueDeserializer =
+        FastSerializerDeserializerFactory.getFastAvroGenericDeserializer(currentValueSchema, currentValueSchema);
+    V currentValue = null;
+    if (recordChangeEvent.currentValue != null) {
+      currentValue = valueDeserializer.deserialize(recordChangeEvent.currentValue.getValue());
+    }
+    V previousValue = null;
+    if (recordChangeEvent.previousValue != null) {
+      previousValue = valueDeserializer.deserialize(recordChangeEvent.previousValue.getValue());
+    }
+    ChangeEvent<V> changeEvent = new ChangeEvent<>(previousValue, currentValue);
+    return new ImmutablePubSubMessage(currentKey, changeEvent, pubSubTopicPartition, offset, timestamp, payloadSize);
+  }
+
+  private void handleControlMessage(ControlMessage controlMessage) {
+    final ControlMessageType controlMessageType = ControlMessageType.valueOf(controlMessage);
+    if (controlMessageType.equals(ControlMessageType.VERSION_SWAP)) {
+      VersionSwap versionSwap = (VersionSwap) controlMessage.controlMessageUnion;
+      LOGGER.info("Obtain version swap message: " + versionSwap);
+      String newServingVersionTopic = versionSwap.newServingVersionTopic.toString();
+      Set<Integer> partitions = new HashSet<>();
+      for (Integer partition: subscribedPartitions) {
+        partitions.add(partition);
+      }
+      unsubscribe(subscribedPartitions);
```

Comment:
I don't know if this is a problem, I'm still wrapping my head around the code a bit, but one thing we MUST guarantee is that once we've received the topic switch message we must not process anymore records from the previous version.  I 'think' this currently runs in serial?  I guess it would depend on how the 'poll' call is getting invoked?  So I'm wondering if we need some kind of synchronization here in order to make the above requirement a guarantee.

Reply:
It is in serial, and you are right, we should not process more records anymore. For synchronization, I think it might be inevitable. Either we use synchronization keyword or having some queue logic inside. But given the sub and unsub are not frequent, and this class is not shared by multiple stores, so I think it should be OK with synchronization keyword.